cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteRuntimeStore.java

```diff
@@ -0,0 +1,127 @@
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.inject.Inject;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+import java.net.HttpURLConnection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Implementation of RuntimeStore, which uses an HTTP Client to execute the actual store operations in a remote
+ * server.
+ */
+public class RemoteRuntimeStore extends RemoteStoreClient implements RuntimeStore {
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(BasicThrowable.class, new BasicThrowableCodec())
+    .registerTypeAdapter(WorkflowTokenDetail.class, new WorkflowTokenDetailCodec())
+    .registerTypeAdapter(WorkflowTokenNodeDetail.class, new WorkflowTokenNodeDetailCodec())
+    .create();
+
+  @Inject
+  public RemoteRuntimeStore(CConfiguration cConf, DiscoveryServiceClient discoveryClient) {
+    super(cConf, discoveryClient);
+  }
+
+  @Override
+  public void compareAndSetStatus(Id.Program id, String pid, ProgramRunStatus expectedStatus,
+                                  ProgramRunStatus updateStatus) {
+    executeRequest("compareAndSetStatus",
+                   id, pid, expectedStatus, updateStatus);
+  }
+
+  @Override
+  public void setStart(Id.Program id, String pid, long startTime, @Nullable String twillRunId,
+                       Map<String, String> runtimeArgs, Map<String, String> systemArgs) {
+    executeRequest("setStart",
+                   id, pid, startTime, twillRunId, runtimeArgs, systemArgs);
+  }
+
+  @Override
+  public void setStop(Id.Program id, String pid, long endTime, ProgramRunStatus runStatus) {
+    // delegates on client side; so corresponding method is not required to be implemented on server side
+    setStop(id, pid, endTime, runStatus, null);
+  }
+
+  @Override
+  public void setStop(Id.Program id, String pid, long endTime, ProgramRunStatus runStatus,
+                      @Nullable BasicThrowable failureCause) {
+    executeRequest("setStop",
+                   id, pid, endTime, runStatus, failureCause);
+  }
+
+  @Override
+  public void setSuspend(Id.Program id, String pid) {
+    executeRequest("setSuspend",
+                   id, pid);
+  }
+
+  @Override
+  public void setResume(Id.Program id, String pid) {
+    executeRequest("setResume",
+                   id, pid);
+  }
+
+  @Override
+  public void updateWorkflowToken(ProgramRunId workflowRunId, WorkflowToken token) {
+    executeRequest("updateWorkflowToken",
+                   workflowRunId, token);
+  }
+
+  @Override
+  public void addWorkflowNodeState(ProgramRunId workflowRunId, WorkflowNodeStateDetail nodeStateDetail) {
+    executeRequest("addWorkflowNodeState",
+                   workflowRunId, nodeStateDetail);
+  }
+
+  private void executeRequest(String methodName, Object... arguments) {
+    doPost("execute/" + methodName, GSON.toJson(createRequest(arguments)));
```

Comment:
It is kind of fatal, so I'm not sure what to do in case of failure.
Perhaps we should retry. @chtyim ?